### PR TITLE
add name mangling

### PR DESCRIPTION
**Pull request by Zamua**
Some discussion on #72

I've added back the test as @hinescd suggested (good idea). I haven't fixed quicksort, nor have I fixed the name collision edge case also pointed out by @hinescd, but this can be a base for others to check out and play with

### CODE REVIEW

**Comment by Zamua** [comment]
Explanation for [why quicksort is failing](https://github.com/JMU-CS/less-java/pull/72#issuecomment-578438979) by @hinescd

**Comment by hinescd** [comment]
Regarding the name collisions issue, the character we use to delimit the parameter types needs to be removed from the set of characters that can make up a class or function name. Suppose we choose T as the delimiter. In the program
```
T {
}
TT {
}
foo(a, b) {
  return a
}
main() {
  foo(T(), TT())
  foo(TT(), T())
}
```
would have a naming collision on ```fooTTTTT(a, b)``` which could return either ```T``` or ```TT```. Removing T as a possible character in the program would solve this issue, but would be an annoyance to the programmer. I think the best candidate for the delimiter is $ because Java allows it in method names, but LJ already doesn't allow it in function or class names.

**Comment by Zamua** [comment]
Ah I see. I went ahead and joined on '$' then. Good idea

**Comment by Zamua** [comment]
Good analysis on the quick sort failure as well. I got that fixed. All tests are passing

**Comment by hinescd** [comment]
Running with
```
T {
}
TT {
}
funT(a) {
  return a
}
fun(a) {
  return a
}
main() {
  funT(T())
  fun(TT())
}
```
Isn't working. I think it might be because there isn't a $ at the beginning of the parameter list in the mangled name, but I'm not sure. For some reason this fails during the java compilation step, but running it with R's instead of T's fails the static analysis.

**Comment by Zamua** [comment]
Okay, I prepended a '$' to the parameterlist

**Comment by hinescd** [comment]
I see the reason it fails with T but not with R. There are too few type inference iterations because LJCheckTypesHaveChanged doesn't recognize that a type changes between HMTypeVar and HMTypeClass("T") since HMType's equals method compares the String representation of the types, which is "T" for both.

**Comment by Zamua** [comment]
Instead of the check I changed HMTypeVar's tostring to be '$T'. Not sure if this is better or worse than your suggestion, but the tests seem to pass, and it involves less code. I figure this is safe because IDs can't contain $ in less-java?

**Comment by hinescd** [comment]
Yeah that looks good to me for now. I think eventually it would be nice to have the HMType classes get proper [equals implementations with inheritance](http://csc.columbusstate.edu/woolbright/java/equals.html), but that's a refactor for another day. Good call adding a new test file for name mangling. It looks like the dotests script is complaining about the test outputs for add and mangled_name_collision though even though their tests are passing. Could you re-run it on those files with the -s flag to set the expected output?
```
./dotests -s tests/add.lj
```

File: src/main/java/com/github/lessjava/visitor/impl/LJInstantiateFunctions.java [hunk 1]
```diff
@@ -190,6 +203,7 @@ public void postVisit(ASTMethodCall node) {
         m = instantiateMethod(m, node.funcCall.arguments);
 
         if (m != null) {
+            // TODO: Mangle method name
```
Comment (lam2mo) [inline src/main/java/com/github/lessjava/visitor/impl/LJInstantiateFunctions.java]: Is this still a TODO? Perhaps we should add an issue before we merge/close this pull request.

Comment (Zamua) [reply src/main/java/com/github/lessjava/visitor/impl/LJInstantiateFunctions.java]: Just pushed, I implemented it for methods, and added a test `test/adder.lj`

**Comment by Zamua** [comment]
I agree we should consider refactoring the equals method @hinescd. I agree we should probably refactor some of the hardcoding @lam2mo, but I also couldn't come up with anything in the moment
